src/solver.rs

```diff
@@ -0,0 +1,134 @@
+//! Solvers for detecting known script kinds.
+
+use bounded_vec::BoundedVec;
+
+use crate::{
+    external::pubkey::PubKey,
+    op,
+    opcode::{push_value::LargeValue::PushdataBytelength, Evaluable, PushValue},
+    script, Opcode,
+};
+
+/// Parses a `script::PubKey` and detects standard scripts.
+///
+/// If successful, returns the script kind, and any relevant properties parsed from the
+/// script. For example, for a P2SH script, the result will contain the script hash; for
+/// P2PKH it will contain the key hash, etc.
+///
+/// Returns `None` if the script is non-standard.
+pub fn standard(script_code: &script::PubKey) -> Option<ScriptKind> {
```

Comment:
To @nuttycom’s comment before, this could be `script_code: &script::Component<Opcode>`. More verbose, but perhaps more correct, since this could also be a redeem script (`script::Redeem` – which all have the same underlying type).

I don’t have a preference, though.